// @dapr/dapr/utils/Client.util.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseEndpoint = exports.getClientOptions = exports.getBulkPublishResponse = exports.getBulkPublishEntries = exports.getContentType = exports.createConfigurationType = exports.getStateConcurrencyValue = exports.getStateConsistencyValue = exports.createHTTPQueryParam = exports.addMetadataToMap = void 0;
const crypto_1 = require("crypto");
const CloudEvent_util_1 = require("./CloudEvent.util");
const CommunicationProtocol_enum_1 = __importDefault(require("../enum/CommunicationProtocol.enum"));
const Settings_util_1 = require("./Settings.util");
const StateConsistency_enum_1 = require("../enum/StateConsistency.enum");
const StateConcurrency_enum_1 = require("../enum/StateConcurrency.enum");
const url_1 = require("url");
/**
 * Adds metadata to a map.
 * @param map Input map
 * @param metadata key value pair of metadata
 */
function addMetadataToMap(map, metadata = {}) {
    for (const [key, value] of Object.entries(metadata)) {
        map.set(key, value);
    }
}
exports.addMetadataToMap = addMetadataToMap;
/**
 * Converts one or multiple sets of data to a querystring
 * Each set of data contains a set of KeyValue Pair
 * An optional "metadata" type can be added in each set, in which case
 * the QS key of each data in the set will be prefixed with "metadata.".
 *
 * Note, the returned value does not contain the "?" prefix.
 *
 * @param params one of multiple set of data
 * @returns HTTP query parameter string
 */
function createHTTPQueryParam(...params) {
    const queryBuilder = new url_1.URLSearchParams();
    for (const group of params) {
        if (!(group === null || group === void 0 ? void 0 : group.data)) {
            continue;
        }
        for (const [key, value] of Object.entries(group.data)) {
            let propName = key;
            if ((group === null || group === void 0 ? void 0 : group.type) === "metadata") {
                propName = `metadata.${propName}`;
            }
            if (value !== undefined) {
                queryBuilder.set(propName, value);
            }
        }
    }
    return queryBuilder.toString();
}
exports.createHTTPQueryParam = createHTTPQueryParam;
/**
 * Return the string representation of a valid consistency configuration
 * @param c
 */
function getStateConsistencyValue(c) {
    switch (c) {
        case StateConsistency_enum_1.StateConsistencyEnum.CONSISTENCY_EVENTUAL:
            return "eventual";
        case StateConsistency_enum_1.StateConsistencyEnum.CONSISTENCY_STRONG:
            return "strong";
        default:
            return undefined;
    }
}
exports.getStateConsistencyValue = getStateConsistencyValue;
/**
 * Return the string representation of a valid concurrency configuration
 * @param c
 */
function getStateConcurrencyValue(c) {
    switch (c) {
        case StateConcurrency_enum_1.StateConcurrencyEnum.CONCURRENCY_FIRST_WRITE:
            return "first-write";
        case StateConcurrency_enum_1.StateConcurrencyEnum.CONCURRENCY_LAST_WRITE:
            return "last-write";
        default:
            return undefined;
    }
}
exports.getStateConcurrencyValue = getStateConcurrencyValue;
/**
 * Converts a Map<string, common_pb.ConfigurationItemProto> to a ConfigurationType object.
 * @param Map<string, common_pb.ConfigurationItemProto>
 * @returns ConfigurationType object
 */
function createConfigurationType(configDict) {
    const configMap = {};
    configDict.forEach(function (v, k) {
        const item = {
            key: k,
            value: v.getValue(),
            version: v.getVersion(),
            metadata: v
                .getMetadataMap()
                .toObject()
                .reduce((result, [key, value]) => {
                // @ts-ignore
                result[key] = value;
                return result;
            }, {}),
        };
        configMap[k] = item;
    });
    return configMap;
}
exports.createConfigurationType = createConfigurationType;
/**
 * Gets the Content-Type for the input data.
 *
 * If the data is a valid Cloud Event, the Content-Type is "application/cloudevents+json".
 * If the data is a JSON object, the Content-Type is "application/json".
 * If the data is a string, the Content-Type is "text/plain".
 * Otherwise, the Content-Type is "application/octet-stream".
 *
 * @param data input data
 * @returns Content-Type header value
 */
function getContentType(data) {
    // Identify the exact type of the input data
    const type = getType(data);
    switch (type) {
        case "Array":
        case "Object":
            if ((0, CloudEvent_util_1.isCloudEvent)(data)) {
                return "application/cloudevents+json";
            }
            else {
                return "application/json";
            }
        case "Boolean":
        case "Number":
        case "String":
            return "text/plain";
        // Uint8Array, Int8Array, Buffer, SlowBuffer, Blob, etc.
        default:
            return "application/octet-stream";
    }
}
exports.getContentType = getContentType;
/**
 * Get the entries for bulk publish request.
 * If entryIDs are missing, generate UUIDs for them.
 * If contentTypes are missing, infer them based on the data using {@link getContentType}.
 *
 * @param messages pubsub bulk publish messages
 * @returns configured entries
 */
function getBulkPublishEntries(messages) {
    return messages.map((message) => {
        // If message is a PubSubBulkPublishEntry, use it directly
        if (typeof message !== "string" && "event" in message) {
            return {
                entryID: message.entryID ? message.entryID : (0, crypto_1.randomUUID)(),
                event: message.event,
                contentType: message.contentType ? message.contentType : getContentType(message.event),
                metadata: message.metadata ? message.metadata : {},
            };
        }
        // Otherwise, create a PubSubBulkPublishEntry from the message
        return {
            entryID: (0, crypto_1.randomUUID)(),
            event: message,
            contentType: getContentType(message),
            metadata: {},
        };
    });
}
exports.getBulkPublishEntries = getBulkPublishEntries;
/**
 * Get the response for bulk publish request.
 *
 * @param response bulk publish API response
 * @param entries entries for bulk publish request
 * @param error error from bulk publish request
 * @returns SDK response for bulk publish request
 */
function getBulkPublishResponse(params) {
    var _a;
    if ("error" in params) {
        // The entire request failed. This typically indicates a problem with the request or the connection.
        const failedMessages = params.entries.map((message) => ({ message, error: params.error }));
        return { failedMessages };
    }
    // Some or all of the entries failed to be published.
    return {
        failedMessages: (_a = params.response.failedEntries.flatMap((entry) => {
            const message = params.entries.find((message) => message.entryID === entry.entryID);
            if (!message) {
                return [];
            }
            return { message, error: new Error(entry.error) };
        })) !== null && _a !== void 0 ? _a : [],
    };
}
exports.getBulkPublishResponse = getBulkPublishResponse;
/**
 * Determine the type of the input data by checking the constructor name
 * If no name is found, we return the typeof the input data
 *
 * @param arr
 * @returns string The Data Type, e.g., Uint8Array, Int8Array, Buffer, SlowBuffer, Blob, Oject, Array, etc.
 */
function getType(o) {
    // If the type is set in the constructor name, return the name
    // e.g., Uint8Array, Int8Array, Buffer, SlowBuffer, Blob, etc.
    if (o.constructor.name) {
        return o.constructor.name;
    }
    return typeof o;
}
/**
 * Prepares DaprClientOptions for use by the DaprClient/DaprServer.
 * If the user does not provide a value for a mandatory option, the default value is used.
 * @param clientOptions DaprClientOptions
 * @param defaultCommunicationProtocol CommunicationProtocolEnum
 * @param defaultLoggerOptions
 * @returns DaprClientOptions
 */
function getClientOptions(clientOptions, defaultCommunicationProtocol, defaultLoggerOptions) {
    var _a, _b, _c, _d;
    const clientCommunicationProtocol = (_a = clientOptions === null || clientOptions === void 0 ? void 0 : clientOptions.communicationProtocol) !== null && _a !== void 0 ? _a : defaultCommunicationProtocol;
    // We decide the host/port/endpoint here
    let daprEndpoint = "";
    if (clientCommunicationProtocol == CommunicationProtocol_enum_1.default.HTTP) {
        daprEndpoint = Settings_util_1.Settings.getDefaultHttpEndpoint();
    }
    else if (clientCommunicationProtocol == CommunicationProtocol_enum_1.default.GRPC) {
        daprEndpoint = Settings_util_1.Settings.getDefaultGrpcEndpoint();
    }
    let host = Settings_util_1.Settings.getDefaultHost();
    let port = Settings_util_1.Settings.getDefaultPort(clientCommunicationProtocol);
    if ((clientOptions === null || clientOptions === void 0 ? void 0 : clientOptions.daprHost) || (clientOptions === null || clientOptions === void 0 ? void 0 : clientOptions.daprPort)) {
        host = (_b = clientOptions === null || clientOptions === void 0 ? void 0 : clientOptions.daprHost) !== null && _b !== void 0 ? _b : host;
        port = (_c = clientOptions === null || clientOptions === void 0 ? void 0 : clientOptions.daprPort) !== null && _c !== void 0 ? _c : port;
    }
    else if (daprEndpoint != "") {
        const [scheme, fqdn, p] = parseEndpoint(daprEndpoint);
        host = `${scheme}://${fqdn}`;
        port = p;
    }
    return {
        daprHost: host,
        daprPort: port,
        communicationProtocol: clientCommunicationProtocol,
        isKeepAlive: clientOptions === null || clientOptions === void 0 ? void 0 : clientOptions.isKeepAlive,
        logger: (_d = clientOptions === null || clientOptions === void 0 ? void 0 : clientOptions.logger) !== null && _d !== void 0 ? _d : defaultLoggerOptions,
        actor: clientOptions === null || clientOptions === void 0 ? void 0 : clientOptions.actor,
        daprApiToken: clientOptions === null || clientOptions === void 0 ? void 0 : clientOptions.daprApiToken,
        maxBodySizeMb: clientOptions === null || clientOptions === void 0 ? void 0 : clientOptions.maxBodySizeMb,
    };
}
exports.getClientOptions = getClientOptions;
/**
 * Parses an endpoint to scheme, fqdn and port
 * Examples:
 *  - http://localhost:3500 -> [http, localhost, 3500]
 *  - localhost:3500 -> [http, localhost, 3500]
 *  - :3500 -> [http, localhost, 3500]
 *  - localhost -> [http, localhost, 80]
 *  - https://localhost:3500 -> [https, localhost, 3500]
 *  - [::1]:3500 -> [http, ::1, 3500]
 *  - [::1] -> [http, ::1, 80]
 *  - http://[2001:db8:1f70:0:999:de8:7648:6e8]:5000 -> [http, 2001:db8:1f70:0:999:de8:7648:6e8, 5000]
 * @throws Error if the address is invalid
 * @param address Endpoint address
 * @returns EndpointTuple (scheme, fqdn, port)
 */
function parseEndpoint(address) {
    // Prefix with a scheme and host when they're not present,
    // because the URL library won't parse it otherwise
    if (address.startsWith(":")) {
        address = "http://localhost" + address;
    }
    if (!address.includes("://")) {
        address = "http://" + address;
    }
    let scheme, fqdn, port;
    try {
        const myURL = new url_1.URL(address);
        scheme = myURL.protocol.replace(":", "");
        fqdn = myURL.hostname.replace("[", "");
        fqdn = fqdn.replace("]", "");
        port = myURL.port || (myURL.protocol == "https:" ? "443" : "80");
    }
    catch (error) {
        throw new Error(`Invalid address: ${address}`);
    }
    return [scheme, fqdn, port];
}
exports.parseEndpoint = parseEndpoint;
