// @dapr/dapr/implementation/Client/HTTPClient/lock.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const UnlockResponse_1 = require("../../../types/lock/UnlockResponse");
class HTTPClientLock {
    constructor(client) {
        this.client = client;
    }
    lock(storeName, resourceId, lockOwner, expiryInSeconds) {
        return __awaiter(this, void 0, void 0, function* () {
            const data = {
                resourceId,
                lockOwner,
                expiryInSeconds,
            };
            const result = yield this.client.executeWithApiVersion("v1.0-alpha1", `/lock/${storeName}`, {
                method: "POST",
                body: data,
            });
            return {
                success: result["success"],
            };
        });
    }
    unlock(storeName, resourceId, lockOwner) {
        return __awaiter(this, void 0, void 0, function* () {
            const data = {
                resourceId,
                lockOwner,
            };
            const result = yield this.client.executeWithApiVersion("v1.0-alpha1", `/unlock/${storeName}`, {
                method: "POST",
                body: data,
            });
            return {
                status: this._statusToLockStatus(result["status"]),
            };
        });
    }
    _statusToLockStatus(status) {
        switch (status) {
            case 0:
                return UnlockResponse_1.LockStatus.Success;
            case 1:
                return UnlockResponse_1.LockStatus.LockDoesNotExist;
            case 2:
                return UnlockResponse_1.LockStatus.LockBelongsToOthers;
            case 3:
            default:
                return UnlockResponse_1.LockStatus.InternalError;
        }
    }
}
exports.default = HTTPClientLock;
